File: code/ex92.py
class Solution:
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        
        #If m = n, no modification need to be done
        if m == n:
            return head
        
        recorder = []
        counter = 1
        
        #Traverse the chain list and get the reverse part
        p = head
        
        while p != None:
            if counter >= m:
                recorder.append(p.val)
            
            counter += 1
            
            if counter > n:
                break
                
            p = p.next
        
        #Record the end part behind reverse part
        end = p.next
            
        recorder.reverse()
        
        
        p = head
        counter = 1
        
        #Reach the insert position
        while p != None:
            counter += 1
            
            if counter >= m:
                break
                
            p = p.next
        
        #When m = 1, head node should be modified
        if m == 1:
            q = ListNode(recorder[0])
            q.next = None
            head = q
            
            p = head
            i = 1
        else:
            i = 0
        
        #Insert the reverse part and link
        while i < len(recorder):
            q = ListNode(recorder[i])
            p.next = q
            p = q
            
            i += 1
        
        #Append the end part
        p.next = end
        
        return head
                

File: code/ex92.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        
        ListNode* p = dummy;
        for (int i = 0; i < m - 1; i++)    p = p->next;
        
        // Reverse the interval [m, n]
        ListNode* h = NULL;
        ListNode* t = NULL;
        ListNode* q = p->next;
        for (int i = 0; i < n - m + 1; i++) {
            ListNode* nxt = q->next;
            q->next = h;
            h = q;
            if (t == NULL)    t = h;
            q = nxt;
        }
        
        t->next = q;
        p->next = h;
        return dummy->next;
    }
};